<!DOCTYPE html>
<html class="writer-html5" lang="zh-CN" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Application with new data &mdash; PROTOCOLS_PYSODB 1.0 文档</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/translations.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="索引" href="../genindex.html" />
    <link rel="search" title="搜索" href="../search.html" />
    <link rel="next" title="Spatial data integration" href="../Spatial%20data%20integration/Spatial%20data%20integration.html" />
    <link rel="prev" title="Reproducibility with original data (DLPFC)" href="Reproducibility%20with%20original%20data%20%28DLPFC%29.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            PROTOCOLS_PYSODB
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="搜索文档" aria-label="搜索文档" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="导航菜单">
              <p class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../SOView/SOView.html">SOView</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Spatially%20variable%20gene%20detection/Spatially%20variable%20gene%20detection.html">Spatially variable gene detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Spatial%20clustering/Spatial%20clustering.html">Spatial clustering</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="Pseudo-spatiotemporal%20analysis.html">Pseudo-spatiotemporal analysis</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Installation.html">Installation</a></li>
<li class="toctree-l2"><a class="reference internal" href="Reproducibility%20with%20original%20data%20%28seqFISH%29.html">Reproducibility with original data (seqFISH)</a></li>
<li class="toctree-l2"><a class="reference internal" href="Reproducibility%20with%20original%20data%20%28DLPFC%29.html">Reproducibility with original data (DLPFC)</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Application with new data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-packages-and-set-configurations">Import packages and set configurations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Streamline-development-of-loading-spatial-data-with-Pysodb">Streamline development of loading spatial data with Pysodb</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Perform-SpaceFlow-for-pseudo-spatiotemporal-analysis">Perform SpaceFlow for pseudo-spatiotemporal analysis</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Spatial%20data%20integration/Spatial%20data%20integration.html">Spatial data integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Spatial%20data%20alignment/Spatial%20data%20alignment.html">Spatial data alignment</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Spatial%20spot%20deconvolution/Spatial%20spot%20deconvolution.html">Spatial spot deconvolution</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="移动版导航菜单" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PROTOCOLS_PYSODB</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="页面导航">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="Pseudo-spatiotemporal%20analysis.html">Pseudo-spatiotemporal analysis</a></li>
      <li class="breadcrumb-item active">Application with new data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/Pseudo-spatiotemporal analysis/Application with new data.ipynb.txt" rel="nofollow"> 查看页面源码</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Application-with-new-data">
<h1>Application with new data<a class="headerlink" href="#Application-with-new-data" title="此标题的永久链接"></a></h1>
<p>This tutorial demonstrates how to pseudo-spatiotemporal analysis on BaristaSeq mouse visual cortex data using Pysodb and SpaceFlow.</p>
<p>A reference paper can be found at <a class="reference external" href="https://www.nature.com/articles/s41467-022-31739-w">https://www.nature.com/articles/s41467-022-31739-w</a>.</p>
<p>This tutorial refers to the following tutorial at <a class="reference external" href="https://github.com/hongleir/SpaceFlow/blob/master/tutorials/seqfish_mouse_embryogenesis.ipynb">https://github.com/hongleir/SpaceFlow/blob/master/tutorials/seqfish_mouse_embryogenesis.ipynb</a>. At the same time, the way of loadding data is modified by using Pysodb.</p>
<section id="Import-packages-and-set-configurations">
<h2>Import packages and set configurations<a class="headerlink" href="#Import-packages-and-set-configurations" title="此标题的永久链接"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the Python warnings module to filter and ignore any warnings that may occur in the program after this point.</span>
<span class="kn">import</span> <span class="nn">warnings</span>
<span class="n">warnings</span><span class="o">.</span><span class="n">filterwarnings</span><span class="p">(</span><span class="s2">&quot;ignore&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import several python packages commonly used in data analysis and visualization.</span>
<span class="c1"># numpy (imported as np) is a package for numerical computing with arrays.</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="c1"># scanpy (imported as sc) is a package for single-cell RNA sequencing analysis.</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="c1"># matplotlib.pyplot (imported as plt) is a package for data visualization.</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="c1"># seaborn (imported as sns) is a package for statistical data visualization, providing high-level interfaces for creating informative and attractive visualizations.</span>
<span class="kn">import</span> <span class="nn">seaborn</span> <span class="k">as</span> <span class="nn">sns</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># from SpaceFlow package import SpaceFlow module</span>
<span class="kn">from</span> <span class="nn">SpaceFlow</span> <span class="kn">import</span> <span class="n">SpaceFlow</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[4]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Imports a palettable package</span>
<span class="kn">import</span> <span class="nn">palettable</span>
<span class="c1"># Create three variables with lists of colors for categorical visualizations and biotechnology-related visualizations, respectively.</span>
<span class="n">cmp_pspace</span> <span class="o">=</span> <span class="n">palettable</span><span class="o">.</span><span class="n">cartocolors</span><span class="o">.</span><span class="n">diverging</span><span class="o">.</span><span class="n">TealRose_7</span><span class="o">.</span><span class="n">mpl_colormap</span>
<span class="n">cmp_domain</span> <span class="o">=</span> <span class="n">palettable</span><span class="o">.</span><span class="n">cartocolors</span><span class="o">.</span><span class="n">qualitative</span><span class="o">.</span><span class="n">Pastel_10</span><span class="o">.</span><span class="n">mpl_colors</span>
<span class="n">cmp_ct</span> <span class="o">=</span> <span class="n">palettable</span><span class="o">.</span><span class="n">cartocolors</span><span class="o">.</span><span class="n">qualitative</span><span class="o">.</span><span class="n">Safe_10</span><span class="o">.</span><span class="n">mpl_colors</span>
</pre></div>
</div>
</div>
</section>
<section id="Streamline-development-of-loading-spatial-data-with-Pysodb">
<h2>Streamline development of loading spatial data with Pysodb<a class="headerlink" href="#Streamline-development-of-loading-spatial-data-with-Pysodb" title="此标题的永久链接"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[5]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import pysodb package</span>
<span class="c1"># Pysodb is a Python package that provides a set of tools for working with SODB databases.</span>
<span class="c1"># SODB is a format used to store data in memory-mapped files for efficient access and querying.</span>
<span class="c1"># This package allows users to interact with SODB files using Python.</span>
<span class="kn">import</span> <span class="nn">pysodb</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[6]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialize the sodb object</span>
<span class="n">sodb</span> <span class="o">=</span> <span class="n">pysodb</span><span class="o">.</span><span class="n">SODB</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[7]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define names of the dataset_name and experiment_name</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s1">&#39;Sun2021Integrating&#39;</span>
<span class="n">experiment_name</span> <span class="o">=</span> <span class="s1">&#39;Slice_1&#39;</span>
<span class="c1"># Load a specific experiment</span>
<span class="c1"># It takes two arguments: the name of the dataset and the name of the experiment to load.</span>
<span class="c1"># Two arguments are available at https://gene.ai.tencent.com/SpatialOmics/.</span>
<span class="c1">#%%time</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">sodb</span><span class="o">.</span><span class="n">load_experiment</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span><span class="n">experiment_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
load experiment[Slice_1] in dataset[Sun2021Integrating]
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[8]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Remove cells belong to the layers &#39;outside_VISp&#39; and &#39;VISp&#39;</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;layer&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;outside_VISp&#39;</span><span class="p">]</span>
<span class="n">adata</span> <span class="o">=</span> <span class="n">adata</span><span class="p">[</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;layer&#39;</span><span class="p">]</span><span class="o">!=</span><span class="s1">&#39;VISp&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[9]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Filter out genes</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">filter_genes</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">min_cells</span><span class="o">=</span><span class="mi">3</span><span class="p">)</span>
</pre></div>
</div>
</div>
</section>
<section id="Perform-SpaceFlow-for-pseudo-spatiotemporal-analysis">
<h2>Perform SpaceFlow for pseudo-spatiotemporal analysis<a class="headerlink" href="#Perform-SpaceFlow-for-pseudo-spatiotemporal-analysis" title="此标题的永久链接"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[10]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create SpaceFlow Object</span>
<span class="c1">#%%time</span>
<span class="c1">#sf = SpaceFlow.SpaceFlow(adata=adata)</span>
<span class="n">sf</span> <span class="o">=</span> <span class="n">SpaceFlow</span><span class="o">.</span><span class="n">SpaceFlow</span><span class="p">(</span>
    <span class="n">count_matrix</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">X</span><span class="p">,</span>
    <span class="n">spatial_locs</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;spatial&#39;</span><span class="p">],</span>
    <span class="n">sample_names</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">obs_names</span><span class="p">,</span>
    <span class="n">gene_names</span><span class="o">=</span><span class="n">adata</span><span class="o">.</span><span class="n">var_names</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Preprocess data</span>
<span class="c1">#%%time</span>
<span class="n">sf</span><span class="o">.</span><span class="n">preprocessing_data</span><span class="p">(</span><span class="n">n_top_genes</span><span class="o">=</span><span class="mi">3000</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Train a deep graph network model</span>
<span class="c1">#%%time</span>
<span class="n">sf</span><span class="o">.</span><span class="n">train</span><span class="p">(</span>
    <span class="n">spatial_regularization_strength</span><span class="o">=</span><span class="mf">0.1</span><span class="p">,</span>
    <span class="n">z_dim</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">lr</span><span class="o">=</span><span class="mf">1e-3</span><span class="p">,</span>
    <span class="n">epochs</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">max_patience</span><span class="o">=</span><span class="mi">50</span><span class="p">,</span>
    <span class="n">min_stop</span><span class="o">=</span><span class="mi">100</span><span class="p">,</span>
    <span class="n">random_seed</span><span class="o">=</span><span class="mi">42</span><span class="p">,</span>
    <span class="n">gpu</span><span class="o">=</span><span class="mi">0</span><span class="p">,</span>
    <span class="n">regularization_acceleration</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span>
    <span class="n">edge_subset_sz</span><span class="o">=</span><span class="mi">1000000</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Epoch 2/1000, Loss: 1.4427732229232788
Epoch 12/1000, Loss: 1.404854416847229
Epoch 22/1000, Loss: 1.355185866355896
Epoch 32/1000, Loss: 1.278242826461792
Epoch 42/1000, Loss: 1.1435610055923462
Epoch 52/1000, Loss: 0.9432040452957153
Epoch 62/1000, Loss: 0.714905321598053
Epoch 72/1000, Loss: 0.5822354555130005
Epoch 82/1000, Loss: 0.5121979713439941
Epoch 92/1000, Loss: 0.425785630941391
Epoch 102/1000, Loss: 0.37741926312446594
Epoch 112/1000, Loss: 0.3343759775161743
Epoch 122/1000, Loss: 0.3119865655899048
Epoch 132/1000, Loss: 0.26779788732528687
Epoch 142/1000, Loss: 0.22297403216362
Epoch 152/1000, Loss: 0.28504329919815063
Epoch 162/1000, Loss: 0.22740697860717773
Epoch 172/1000, Loss: 0.234877809882164
Epoch 182/1000, Loss: 0.1949552297592163
Epoch 192/1000, Loss: 0.20016708970069885
Epoch 202/1000, Loss: 0.2028239667415619
Epoch 212/1000, Loss: 0.173057422041893
Epoch 222/1000, Loss: 0.21973812580108643
Epoch 232/1000, Loss: 0.17184185981750488
Epoch 242/1000, Loss: 0.2074703425168991
Epoch 252/1000, Loss: 0.19310833513736725
Epoch 262/1000, Loss: 0.2128731906414032
Epoch 272/1000, Loss: 0.17560149729251862
Epoch 282/1000, Loss: 0.2080163210630417
Epoch 292/1000, Loss: 0.18244342505931854
Epoch 302/1000, Loss: 0.17130610346794128
Epoch 312/1000, Loss: 0.16225826740264893
Epoch 322/1000, Loss: 0.15506717562675476
Epoch 332/1000, Loss: 0.1312013417482376
Epoch 342/1000, Loss: 0.14738863706588745
Epoch 352/1000, Loss: 0.1790708303451538
Epoch 362/1000, Loss: 0.1254740208387375
Epoch 372/1000, Loss: 0.1862727850675583
Epoch 382/1000, Loss: 0.17113451659679413
Epoch 392/1000, Loss: 0.141239196062088
Epoch 402/1000, Loss: 0.11042129248380661
Epoch 412/1000, Loss: 0.1695185899734497
Epoch 422/1000, Loss: 0.11782366037368774
Epoch 432/1000, Loss: 0.14781805872917175
Epoch 442/1000, Loss: 0.17524565756320953
Epoch 452/1000, Loss: 0.13630664348602295
Epoch 462/1000, Loss: 0.15702477097511292
Epoch 472/1000, Loss: 0.11048941314220428
Epoch 482/1000, Loss: 0.12970110774040222
Epoch 492/1000, Loss: 0.15136636793613434
Epoch 502/1000, Loss: 0.10564137995243073
Epoch 512/1000, Loss: 0.14658908545970917
Epoch 522/1000, Loss: 0.11960890889167786
Epoch 532/1000, Loss: 0.13005761802196503
Epoch 542/1000, Loss: 0.11053520441055298
Epoch 552/1000, Loss: 0.11907373368740082
Epoch 562/1000, Loss: 0.1232338398694992
Epoch 572/1000, Loss: 0.11129128932952881
Epoch 582/1000, Loss: 0.10171715170145035
Epoch 592/1000, Loss: 0.09877597540616989
Epoch 602/1000, Loss: 0.12555311620235443
Epoch 612/1000, Loss: 0.10553622245788574
Epoch 622/1000, Loss: 0.13101576268672943
Epoch 632/1000, Loss: 0.1205601841211319
Training complete!
Embedding is saved at ./embedding.tsv
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
array([[ 0.8899192 ,  0.01435345,  0.7866027 , ...,  0.5507234 ,
        -0.03758647, -0.08572218],
       [ 0.8003285 ,  0.00401396,  0.72124624, ...,  0.42994535,
        -0.04228881, -0.06209074],
       [ 0.8963664 ,  0.01202957,  0.7585074 , ...,  0.59778905,
        -0.03772109, -0.08880965],
       ...,
       [ 0.48202616,  0.01098088,  0.6471445 , ..., -0.00690297,
         0.73529345,  0.32098433],
       [ 0.5201945 ,  0.00505677,  0.6052369 , ..., -0.00182656,
         0.46181548,  0.19144082],
       [ 0.4544619 ,  0.00249925,  0.5253095 , ..., -0.00313081,
         0.6240246 ,  0.4014499 ]], dtype=float32)
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Idenfify the spatiotemporal patterns through pseudo-Spatiotemporal Map (pSM)</span>
<span class="n">sf</span><span class="o">.</span><span class="n">pseudo_Spatiotemporal_Map</span><span class="p">(</span><span class="n">pSM_values_save_filepath</span><span class="o">=</span><span class="s2">&quot;./pSM_values.tsv&quot;</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">resolution</span><span class="o">=</span><span class="mf">1.0</span><span class="p">)</span>
<br/></pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Performing pseudo-Spatiotemporal Map
pseudo-Spatiotemporal Map(pSM) calculation complete, pSM values of cells or spots saved at ./pSM_values.tsv!
</pre></div></div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a new column called &#39;pspace&#39; from pSM values</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;pspace&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">sf</span><span class="o">.</span><span class="n">pSM_values</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[15]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a UMAP projection from SpaceFlow&#39;s embedding</span>
<span class="n">adata</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;embedding&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">sf</span><span class="o">.</span><span class="n">embedding</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pp</span><span class="o">.</span><span class="n">neighbors</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span> <span class="n">n_neighbors</span><span class="o">=</span><span class="mi">20</span><span class="p">,</span> <span class="n">use_rep</span><span class="o">=</span><span class="s1">&#39;embedding&#39;</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Since this dataset contains &#39;depth_um&#39; in obs, a iroot was set to be the cell with smallest depth_um.</span>
<span class="c1"># For datasets without depth information, one can use the pseudo_Spatiotemporal_Map function</span>
<span class="c1"># Here set the iroot according to &quot;depth_um&quot;</span>

<span class="c1"># Select the root cell for trajectory inference based on its depth, by setting the index of the cell with the smallest &#39;depth_um&#39; value as the root</span>
<span class="n">adata</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;iroot&#39;</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">argmin</span><span class="p">(</span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span><span class="p">[</span><span class="s1">&#39;depth_um&#39;</span><span class="p">])</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">diffmap</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">dpt</span><span class="p">(</span><span class="n">adata</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Plot spatial embedding and UMAP embedding for diffusion pseudotime and layer(label), respectively</span>

<span class="c1"># si = &#39;Slice_1&#39;</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s1">&#39;spatial&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dpt_pseudotime&#39;</span><span class="p">],</span><span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">color_map</span><span class="o">=</span><span class="n">cmp_pspace</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_pspace.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_pspace.pdf&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>

<span class="n">ax</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s1">&#39;spatial&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;layer&#39;</span><span class="p">,</span><span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">cmp_domain</span><span class="p">)</span>
<span class="n">ax</span><span class="o">.</span><span class="n">axis</span><span class="p">(</span><span class="s1">&#39;equal&#39;</span><span class="p">)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_domain.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_domain.pdf&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;layer&#39;</span><span class="p">,</span><span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">palette</span><span class="o">=</span><span class="n">cmp_domain</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="c1">#ax.axis(&#39;equal&#39;)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_UMAP_domain.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_UMAP_domain.pdf&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>

<span class="n">fig</span><span class="p">,</span> <span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">()</span>
<span class="n">fig</span><span class="o">.</span><span class="n">set_size_inches</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span> <span class="mi">4</span><span class="p">)</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s1">&#39;X_umap&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s1">&#39;dpt_pseudotime&#39;</span><span class="p">],</span><span class="n">show</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span><span class="n">color_map</span><span class="o">=</span><span class="n">cmp_pspace</span><span class="p">,</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="c1">#ax.axis(&#39;equal&#39;)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_UMAP_pspace.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_UMAP_pspace.pdf&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;dpt_pseudotime&#39;}, xlabel=&#39;X_umap1&#39;, ylabel=&#39;X_umap2&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_1.png" src="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_1.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_2.png" src="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_2.png" />
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_3.png" src="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_3.png" />
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_4.png" src="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_21_4.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check whether the pseudo-spatiotemporal (dpt_pseudotime) value from infer increases according to the layer of the cortex</span>
<span class="n">adata_use</span> <span class="o">=</span> <span class="n">adata</span>
<span class="n">fig</span><span class="p">,</span><span class="n">ax</span> <span class="o">=</span> <span class="n">plt</span><span class="o">.</span><span class="n">subplots</span><span class="p">(</span><span class="n">figsize</span><span class="o">=</span><span class="p">(</span><span class="mi">4</span><span class="p">,</span><span class="mi">4</span><span class="p">))</span>
<span class="n">sns</span><span class="o">.</span><span class="n">violinplot</span><span class="p">(</span><span class="n">data</span><span class="o">=</span><span class="n">adata_use</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span><span class="n">x</span><span class="o">=</span><span class="s1">&#39;layer&#39;</span><span class="p">,</span><span class="n">y</span><span class="o">=</span><span class="s1">&#39;dpt_pseudotime&#39;</span><span class="p">,</span><span class="n">palette</span> <span class="o">=</span> <span class="n">adata_use</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;layer_colors&#39;</span><span class="p">],</span><span class="n">ax</span><span class="o">=</span><span class="n">ax</span><span class="p">)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_violin.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_violin.pdf&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: xlabel=&#39;layer&#39;, ylabel=&#39;dpt_pseudotime&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_22_1.png" src="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_22_1.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata</span><span class="o">.</span><span class="n">obs</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Slice</th>
      <th>x</th>
      <th>y</th>
      <th>Dist to pia</th>
      <th>Dist to bottom</th>
      <th>Angle</th>
      <th>unused-1</th>
      <th>unused-2</th>
      <th>x_um</th>
      <th>y_um</th>
      <th>depth_um</th>
      <th>layer</th>
      <th>leiden</th>
      <th>pspace</th>
      <th>dpt_pseudotime</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>20022</th>
      <td>1</td>
      <td>12186.20</td>
      <td>8617.70</td>
      <td>1029.630</td>
      <td>205.270</td>
      <td>174.894</td>
      <td>0</td>
      <td>0</td>
      <td>1218.620</td>
      <td>861.770</td>
      <td>1093.752193</td>
      <td>VISp_VI</td>
      <td>1</td>
      <td>0.911279</td>
      <td>0.903298</td>
    </tr>
    <tr>
      <th>20023</th>
      <td>1</td>
      <td>12789.10</td>
      <td>8690.82</td>
      <td>1072.540</td>
      <td>168.011</td>
      <td>172.700</td>
      <td>0</td>
      <td>0</td>
      <td>1278.910</td>
      <td>869.082</td>
      <td>1139.629730</td>
      <td>VISp_VI</td>
      <td>4</td>
      <td>0.892462</td>
      <td>0.882942</td>
    </tr>
    <tr>
      <th>20024</th>
      <td>1</td>
      <td>11927.80</td>
      <td>8715.20</td>
      <td>1003.280</td>
      <td>231.018</td>
      <td>177.522</td>
      <td>0</td>
      <td>0</td>
      <td>1192.780</td>
      <td>871.520</td>
      <td>1066.880345</td>
      <td>VISp_VI</td>
      <td>1</td>
      <td>0.928142</td>
      <td>0.921913</td>
    </tr>
    <tr>
      <th>20025</th>
      <td>1</td>
      <td>12860.60</td>
      <td>8729.82</td>
      <td>1075.760</td>
      <td>165.770</td>
      <td>172.700</td>
      <td>0</td>
      <td>0</td>
      <td>1286.060</td>
      <td>872.982</td>
      <td>1143.373530</td>
      <td>VISp_VI</td>
      <td>4</td>
      <td>0.895546</td>
      <td>0.886348</td>
    </tr>
    <tr>
      <th>20027</th>
      <td>1</td>
      <td>12587.60</td>
      <td>8760.70</td>
      <td>1052.380</td>
      <td>187.131</td>
      <td>172.700</td>
      <td>0</td>
      <td>0</td>
      <td>1258.760</td>
      <td>876.070</td>
      <td>1119.019231</td>
      <td>VISp_VI</td>
      <td>3</td>
      <td>0.909229</td>
      <td>0.901504</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>230184</th>
      <td>1</td>
      <td>3239.60</td>
      <td>8952.13</td>
      <td>307.763</td>
      <td>971.248</td>
      <td>178.930</td>
      <td>0</td>
      <td>0</td>
      <td>323.960</td>
      <td>895.213</td>
      <td>333.538252</td>
      <td>VISp_II/III</td>
      <td>4</td>
      <td>0.715771</td>
      <td>0.520664</td>
    </tr>
    <tr>
      <th>230186</th>
      <td>1</td>
      <td>2713.10</td>
      <td>9006.56</td>
      <td>261.052</td>
      <td>1018.700</td>
      <td>177.579</td>
      <td>0</td>
      <td>0</td>
      <td>271.310</td>
      <td>900.656</td>
      <td>286.855966</td>
      <td>VISp_II/III</td>
      <td>2</td>
      <td>0.703140</td>
      <td>0.509154</td>
    </tr>
    <tr>
      <th>230187</th>
      <td>1</td>
      <td>2193.10</td>
      <td>9009.00</td>
      <td>214.572</td>
      <td>1063.490</td>
      <td>172.257</td>
      <td>0</td>
      <td>0</td>
      <td>219.310</td>
      <td>900.900</td>
      <td>243.621417</td>
      <td>VISp_II/III</td>
      <td>0</td>
      <td>0.783840</td>
      <td>0.606285</td>
    </tr>
    <tr>
      <th>230188</th>
      <td>1</td>
      <td>2405.97</td>
      <td>9015.50</td>
      <td>234.015</td>
      <td>1045.450</td>
      <td>171.641</td>
      <td>0</td>
      <td>0</td>
      <td>240.597</td>
      <td>901.550</td>
      <td>260.900559</td>
      <td>VISp_II/III</td>
      <td>0</td>
      <td>0.727675</td>
      <td>0.530846</td>
    </tr>
    <tr>
      <th>230192</th>
      <td>1</td>
      <td>2963.35</td>
      <td>9161.75</td>
      <td>273.140</td>
      <td>1005.830</td>
      <td>178.946</td>
      <td>0</td>
      <td>0</td>
      <td>296.335</td>
      <td>916.175</td>
      <td>298.913498</td>
      <td>VISp_II/III</td>
      <td>0</td>
      <td>0.720848</td>
      <td>0.534821</td>
    </tr>
  </tbody>
</table>
<p>1525 rows × 15 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_use</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;layer_colors&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
[&#39;#66c5cc&#39;, &#39;#f6cf71&#39;, &#39;#f89c74&#39;, &#39;#dcb0f2&#39;, &#39;#87c55f&#39;, &#39;#9eb9f3&#39;]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Check whether the pseudo-spatiotemporal (dpt_pseudotime) from infer was correlated with their depth_um (depth_um)</span>
<span class="n">adata_use</span> <span class="o">=</span> <span class="n">adata</span>
<span class="n">g</span> <span class="o">=</span> <span class="n">sns</span><span class="o">.</span><span class="n">jointplot</span><span class="p">(</span><span class="n">x</span><span class="o">=</span><span class="s2">&quot;depth_um&quot;</span><span class="p">,</span> <span class="n">y</span><span class="o">=</span><span class="s2">&quot;dpt_pseudotime&quot;</span><span class="p">,</span> <span class="n">data</span><span class="o">=</span><span class="n">adata_use</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span><span class="n">hue</span><span class="o">=</span><span class="s1">&#39;layer&#39;</span><span class="p">,</span>
                  <span class="n">palette</span><span class="o">=</span><span class="nb">list</span><span class="p">(</span><span class="n">adata_use</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s1">&#39;layer_colors&#39;</span><span class="p">]),</span>
                  <span class="c1"># kind=&quot;reg&quot;,</span>
                  <span class="c1"># truncate=False,</span>
                  <span class="c1"># xlim=(0, 60), ylim=(0, 12),</span>
                  <span class="c1"># color=&quot;m&quot;,</span>
                  <span class="n">height</span><span class="o">=</span><span class="mi">7</span><span class="p">)</span>
<span class="n">g</span><span class="o">.</span><span class="n">ax_joint</span><span class="o">.</span><span class="n">legend_</span><span class="o">.</span><span class="n">remove</span><span class="p">()</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_jointplot.png&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
<span class="c1"># plt.savefig(f&#39;../figures/pspace/BaristaSeq_{si}_jointplot.pdf&#39;,bbox_inches=&#39;tight&#39;,transparent=True,dpi=400)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_25_0.png" src="../_images/Pseudo-spatiotemporal_analysis_Application_with_new_data_25_0.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="页脚">
        <a href="Reproducibility%20with%20original%20data%20%28DLPFC%29.html" class="btn btn-neutral float-left" title="Reproducibility with original data (DLPFC)" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> 上一页</a>
        <a href="../Spatial%20data%20integration/Spatial%20data%20integration.html" class="btn btn-neutral float-right" title="Spatial data integration" accesskey="n" rel="next">下一页 <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; 版权所有 2023, LINSENLIN.</p>
  </div>

  利用 <a href="https://www.sphinx-doc.org/">Sphinx</a> 构建，使用的 
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">主题</a>
    由 <a href="https://readthedocs.org">Read the Docs</a> 开发.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>