<!DOCTYPE html>
<html class="writer-html5" lang="zh-CN" >
<head>
  <meta charset="utf-8" /><meta name="generator" content="Docutils 0.18.1: http://docutils.sourceforge.net/" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Reproducibility with original data &mdash; PROTOCOLS_PYSODB 1.0 文档</title>
      <link rel="stylesheet" href="../_static/pygments.css" type="text/css" />
      <link rel="stylesheet" href="../_static/css/theme.css" type="text/css" />
      <link rel="stylesheet" href="../_static/nbsphinx-code-cells.css" type="text/css" />
  <!--[if lt IE 9]>
    <script src="../_static/js/html5shiv.min.js"></script>
  <![endif]-->
  
        <script data-url_root="../" id="documentation_options" src="../_static/documentation_options.js"></script>
        <script src="../_static/doctools.js"></script>
        <script src="../_static/sphinx_highlight.js"></script>
        <script src="../_static/translations.js"></script>
        <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
        <script>window.MathJax = {"tex": {"inlineMath": [["$", "$"], ["\\(", "\\)"]], "processEscapes": true}, "options": {"ignoreHtmlClass": "tex2jax_ignore|mathjax_ignore|document", "processHtmlClass": "tex2jax_process|mathjax_process|math|output_area"}}</script>
        <script defer="defer" src="https://cdn.jsdelivr.net/npm/mathjax@3/es5/tex-mml-chtml.js"></script>
    <script src="../_static/js/theme.js"></script>
    <link rel="index" title="索引" href="../genindex.html" />
    <link rel="search" title="搜索" href="../search.html" />
    <link rel="next" title="Application with new data" href="Application%20with%20new%20data.html" />
    <link rel="prev" title="Installation" href="Installation.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../index.html" class="icon icon-home">
            PROTOCOLS_PYSODB
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="搜索文档" aria-label="搜索文档" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="导航菜单">
              <p class="caption" role="heading"><span class="caption-text">Tutorial:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="../Alignment/Alignment.html">Alignment</a></li>
<li class="toctree-l1 current"><a class="reference internal" href="Deconvolution.html">Deconvolution</a><ul class="current">
<li class="toctree-l2"><a class="reference internal" href="Installation.html">Installation</a></li>
<li class="toctree-l2 current"><a class="current reference internal" href="#">Reproducibility with original data</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#Import-packages-and-set-configurations">Import packages and set configurations</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Load-a-single-cell-dataset">Load a single cell dataset</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Streamline-development-of-loading-spatial-data-with-SOBD">Streamline development of loading spatial data with SOBD</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Preparation">Preparation</a></li>
<li class="toctree-l3"><a class="reference internal" href="#Perform-Tangram-for-alignment">Perform Tangram for alignment</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="Application%20with%20new%20data.html">Application with new data</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="../Integration/Integration.html">Integration</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Pseudo-spatiotemporal%20analysis/Pseudo-spatiotemporal%20analysis.html">Pseudo-spatiotemporal analysis</a></li>
<li class="toctree-l1"><a class="reference internal" href="../SOView/SOView.html">SOView</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Spatial%20clustering/Spatial%20clustering.html">Spatial clustering</a></li>
<li class="toctree-l1"><a class="reference internal" href="../Spatially%20variable%20gene/Spatially%20variable%20gene.html">Spatially variable gene</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="移动版导航菜单" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../index.html">PROTOCOLS_PYSODB</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="页面导航">
  <ul class="wy-breadcrumbs">
      <li><a href="../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="Deconvolution.html">Deconvolution</a></li>
      <li class="breadcrumb-item active">Reproducibility with original data</li>
      <li class="wy-breadcrumbs-aside">
            <a href="../_sources/Deconvolution/Reproducibility with original data.ipynb.txt" rel="nofollow"> 查看页面源码</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="Reproducibility-with-original-data">
<h1>Reproducibility with original data<a class="headerlink" href="#Reproducibility-with-original-data" title="此标题的永久链接"></a></h1>
<p>This tutorial demonstrates deconvolution on 10x Visium mouse brain data using SODB and Tangram.</p>
<p>A reference paper can be found at <a class="reference external" href="https://www.nature.com/articles/s41592-021-01264-7">https://www.nature.com/articles/s41592-021-01264-7</a>.</p>
<p>This tutorial refers to the following tutorial at <a class="reference external" href="https://squidpy.readthedocs.io/en/stable/external_tutorials/tutorial_tangram.html">https://squidpy.readthedocs.io/en/stable/external_tutorials/tutorial_tangram.html</a>. At the same time, the way of loadding data is modified by using SODB.</p>
<p>（需要增加单细胞数据链接）</p>
<section id="Import-packages-and-set-configurations">
<h2>Import packages and set configurations<a class="headerlink" href="#Import-packages-and-set-configurations" title="此标题的永久链接"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[1]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import several Python packages, including:</span>
<span class="c1"># scanpy: a Python package for single-cell RNA sequencing analysis</span>
<span class="kn">import</span> <span class="nn">scanpy</span> <span class="k">as</span> <span class="nn">sc</span>
<span class="c1"># squidpy: a Python package for spatial transcriptomics analysis</span>
<span class="kn">import</span> <span class="nn">squidpy</span> <span class="k">as</span> <span class="nn">sq</span>
<span class="c1"># numpy: a Python package for scientific computing with arrays</span>
<span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="c1"># pandas: a Python package for data manipulation and analysis</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>
<span class="c1"># anndata: a Python package for handling annotated data objects in genomics</span>
<span class="kn">import</span> <span class="nn">anndata</span> <span class="k">as</span> <span class="nn">ad</span>
<span class="c1"># pathlib: a Python module for working with file system paths</span>
<span class="kn">import</span> <span class="nn">pathlib</span>
<span class="c1"># matplotlib: a Python plotting library</span>
<span class="kn">import</span> <span class="nn">matplotlib.pyplot</span> <span class="k">as</span> <span class="nn">plt</span>
<span class="kn">import</span> <span class="nn">matplotlib</span> <span class="k">as</span> <span class="nn">mpl</span>
<span class="c1"># skimage: a Python package for image processing</span>
<span class="kn">import</span> <span class="nn">skimage</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[2]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import tangram for spatial deconvolution</span>
<span class="kn">import</span> <span class="nn">tangram</span> <span class="k">as</span> <span class="nn">tg</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[3]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># print a header message, and the version of the squidpy and tangram packages</span>
<span class="n">sc</span><span class="o">.</span><span class="n">logging</span><span class="o">.</span><span class="n">print_header</span><span class="p">()</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;squidpy==</span><span class="si">{</span><span class="n">sq</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
<span class="nb">print</span><span class="p">(</span><span class="sa">f</span><span class="s2">&quot;tangram==</span><span class="si">{</span><span class="n">tg</span><span class="o">.</span><span class="n">__version__</span><span class="si">}</span><span class="s2">&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
scanpy==1.9.3 anndata==0.8.0 umap==0.5.3 numpy==1.22.4 scipy==1.9.1 pandas==1.5.3 scikit-learn==1.2.2 statsmodels==0.13.5 python-igraph==0.10.4 pynndescent==0.5.8
squidpy==1.2.3
tangram==1.0.4
</pre></div></div>
</div>
</section>
<section id="Load-a-single-cell-dataset">
<h2>Load a single cell dataset<a class="headerlink" href="#Load-a-single-cell-dataset" title="此标题的永久链接"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[11]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Load the reference single cell dataset</span>
<span class="c1"># The input sc data has been normalized and log-transformed</span>
<span class="n">adata_sc</span> <span class="o">=</span> <span class="n">sc</span><span class="o">.</span><span class="n">read_h5ad</span><span class="p">(</span><span class="s1">&#39;data/Visium/sc_mouse_cortex.h5ad&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Print out the metadata of adata_sc</span>
<span class="n">adata_sc</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[12]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 21697 × 36826
    obs: &#39;sample_name&#39;, &#39;organism&#39;, &#39;donor_sex&#39;, &#39;cell_class&#39;, &#39;cell_subclass&#39;, &#39;cell_cluster&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_mt&#39;, &#39;log1p_total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;n_counts&#39;
    var: &#39;mt&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;
    uns: &#39;cell_class_colors&#39;, &#39;cell_subclass_colors&#39;, &#39;hvg&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize a UMAP projection colored by cell_subclass</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">umap</span><span class="p">(</span>
    <span class="n">adata_sc</span><span class="p">,</span> <span class="n">color</span><span class="o">=</span><span class="s2">&quot;cell_subclass&quot;</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="mi">10</span><span class="p">,</span> <span class="n">frameon</span><span class="o">=</span><span class="kc">False</span><span class="p">,</span> <span class="n">show</span><span class="o">=</span><span class="kc">False</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[13]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
&lt;Axes: title={&#39;center&#39;: &#39;cell_subclass&#39;}, xlabel=&#39;UMAP1&#39;, ylabel=&#39;UMAP2&#39;&gt;
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Deconvolution_Reproducibility_with_original_data_10_1.png" src="../_images/Deconvolution_Reproducibility_with_original_data_10_1.png" />
</div>
</div>
</section>
<section id="Streamline-development-of-loading-spatial-data-with-SOBD">
<h2>Streamline development of loading spatial data with SOBD<a class="headerlink" href="#Streamline-development-of-loading-spatial-data-with-SOBD" title="此标题的永久链接"></a></h2>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[14]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Import pysodb package</span>
<span class="c1"># Pysodb is a Python package that provides a set of tools for working with SODB databases.</span>
<span class="c1"># SODB is a format used to store data in memory-mapped files for efficient access and querying.</span>
<span class="c1"># This package allows users to interact with SODB files using Python.</span>
<span class="kn">import</span> <span class="nn">pysodb</span>
</pre></div>
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[16]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Initialization</span>
<span class="n">sodb</span> <span class="o">=</span> <span class="n">pysodb</span><span class="o">.</span><span class="n">SODB</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[17]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Define the name of the dataset_name and experiment_name</span>
<span class="n">dataset_name</span> <span class="o">=</span> <span class="s1">&#39;Biancalani2021Deep&#39;</span>
<span class="n">experiment_name</span> <span class="o">=</span> <span class="s1">&#39;visium_fluo_crop&#39;</span>
<span class="c1"># Load a specific experiment</span>
<span class="c1"># It takes two arguments: the name of the dataset and the name of the experiment to load.</span>
<span class="c1"># Two arguments are available at https://gene.ai.tencent.com/SpatialOmics/.</span>
<span class="n">adata_st</span> <span class="o">=</span> <span class="n">sodb</span><span class="o">.</span><span class="n">load_experiment</span><span class="p">(</span><span class="n">dataset_name</span><span class="p">,</span><span class="n">experiment_name</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
load experiment[visium_fluo_crop] in dataset[Biancalani2021Deep]
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">adata_st</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[18]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 704 × 16562
    obs: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_MT&#39;, &#39;log1p_total_counts_MT&#39;, &#39;pct_counts_MT&#39;, &#39;n_counts&#39;, &#39;leiden&#39;, &#39;cluster&#39;
    var: &#39;gene_ids&#39;, &#39;feature_types&#39;, &#39;genome&#39;, &#39;MT&#39;, &#39;n_cells_by_counts&#39;, &#39;mean_counts&#39;, &#39;log1p_mean_counts&#39;, &#39;pct_dropout_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;n_cells&#39;, &#39;highly_variable&#39;, &#39;highly_variable_rank&#39;, &#39;means&#39;, &#39;variances&#39;, &#39;variances_norm&#39;, &#39;dispersions&#39;, &#39;dispersions_norm&#39;
    uns: &#39;cluster_colors&#39;, &#39;hvg&#39;, &#39;leiden&#39;, &#39;leiden_colors&#39;, &#39;log1p&#39;, &#39;moranI&#39;, &#39;neighbors&#39;, &#39;pca&#39;, &#39;spatial&#39;, &#39;spatial_neighbors&#39;, &#39;umap&#39;
    obsm: &#39;X_pca&#39;, &#39;X_umap&#39;, &#39;spatial&#39;
    varm: &#39;PCs&#39;
    obsp: &#39;connectivities&#39;, &#39;distances&#39;, &#39;spatial_connectivities&#39;, &#39;spatial_distances&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[19]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a spatial scatter plot colored by cluster label</span>
<span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span><span class="n">adata_st</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Deconvolution_Reproducibility_with_original_data_16_0.png" src="../_images/Deconvolution_Reproducibility_with_original_data_16_0.png" />
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[20]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize embedding base on &#39;spatial&#39; with points colored by &#39;cluster&#39; label</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_st</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s1">&#39;spatial&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Deconvolution_Reproducibility_with_original_data_17_0.png" src="../_images/Deconvolution_Reproducibility_with_original_data_17_0.png" />
</div>
</div>
<div class="nbinput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[21]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># select a subset based on the &quot;Cortex_{i}&quot; of &#39;adata_st.obs.cluster&#39;</span>
<span class="c1"># the arange of i is form 0 to 4</span>
<span class="c1"># And creates a copy of the resulting subset</span>

<span class="n">adata_st</span> <span class="o">=</span> <span class="n">adata_st</span><span class="p">[</span>
    <span class="n">adata_st</span><span class="o">.</span><span class="n">obs</span><span class="o">.</span><span class="n">cluster</span><span class="o">.</span><span class="n">isin</span><span class="p">([</span><span class="sa">f</span><span class="s2">&quot;Cortex_</span><span class="si">{</span><span class="n">i</span><span class="si">}</span><span class="s2">&quot;</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="n">np</span><span class="o">.</span><span class="n">arange</span><span class="p">(</span><span class="mi">1</span><span class="p">,</span> <span class="mi">5</span><span class="p">)])</span>
<span class="p">]</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
</pre></div>
</div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[22]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Visualize embedding base on &#39;spatial&#39; with points colored by a new &#39;cluster&#39; label</span>
<span class="n">sc</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">embedding</span><span class="p">(</span><span class="n">adata_st</span><span class="p">,</span><span class="n">basis</span><span class="o">=</span><span class="s1">&#39;spatial&#39;</span><span class="p">,</span><span class="n">color</span><span class="o">=</span><span class="s1">&#39;cluster&#39;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Deconvolution_Reproducibility_with_original_data_19_0.png" src="../_images/Deconvolution_Reproducibility_with_original_data_19_0.png" />
</div>
</div>
</section>
<section id="Preparation">
<h2>Preparation<a class="headerlink" href="#Preparation" title="此标题的永久链接"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[23]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Perform differential gene expression analysis across &#39;cell_subclasses&#39; in &#39;adata_sc&#39;</span>
<span class="n">sc</span><span class="o">.</span><span class="n">tl</span><span class="o">.</span><span class="n">rank_genes_groups</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">groupby</span><span class="o">=</span><span class="s2">&quot;cell_subclass&quot;</span><span class="p">,</span> <span class="n">use_raw</span><span class="o">=</span><span class="kc">False</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
WARNING: Default of the method has been changed to &#39;t-test&#39; from &#39;t-test_overestim_var&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Create a Pandas DataFrame called &quot;markers_df&quot; by extracting the top 100 differentially expressed genes from &#39;adata_sc&#39;</span>
<span class="n">markers_df</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">adata_sc</span><span class="o">.</span><span class="n">uns</span><span class="p">[</span><span class="s2">&quot;rank_genes_groups&quot;</span><span class="p">][</span><span class="s2">&quot;names&quot;</span><span class="p">])</span><span class="o">.</span><span class="n">iloc</span><span class="p">[</span><span class="mi">0</span><span class="p">:</span><span class="mi">100</span><span class="p">,</span> <span class="p">:]</span>
<span class="c1"># Create a NumPy array called &quot;genes_sc&quot; by extracting the unique values from the &quot;value&quot; column of a melted version of the &quot;markers_df&quot;</span>
<span class="n">genes_sc</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">unique</span><span class="p">(</span><span class="n">markers_df</span><span class="o">.</span><span class="n">melt</span><span class="p">()</span><span class="o">.</span><span class="n">value</span><span class="o">.</span><span class="n">values</span><span class="p">)</span>
<span class="c1"># Extracte the names of genes from &quot;adata_st&quot;</span>
<span class="n">genes_st</span> <span class="o">=</span> <span class="n">adata_st</span><span class="o">.</span><span class="n">var_names</span><span class="o">.</span><span class="n">values</span>
<span class="c1"># Create a Python list called &quot;genes&quot;</span>
<span class="c1"># Contain the intersection of genes identified as differentially expressed in  &quot;genes_sc&quot; and genes detected in &quot;genes_st&quot;.</span>
<span class="n">genes</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes_sc</span><span class="p">)</span><span class="o">.</span><span class="n">intersection</span><span class="p">(</span><span class="nb">set</span><span class="p">(</span><span class="n">genes_st</span><span class="p">)))</span>
<span class="c1"># The length of &quot;genes&quot;</span>
<span class="nb">len</span><span class="p">(</span><span class="n">genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[24]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
1281
</pre></div></div>
</div>
</section>
<section id="Perform-Tangram-for-alignment">
<h2>Perform Tangram for alignment<a class="headerlink" href="#Perform-Tangram-for-alignment" title="此标题的永久链接"></a></h2>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[25]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the Tangram to align the gene expression profiles of &quot;adata_sc&quot; and &quot;adata_st&quot; based on the shared set of genes identified by the intersection of &quot;genes_sc&quot; and &quot;genes_st&quot;.</span>
<span class="n">tg</span><span class="o">.</span><span class="n">pp_adatas</span><span class="p">(</span><span class="n">adata_sc</span><span class="p">,</span> <span class="n">adata_st</span><span class="p">,</span> <span class="n">genes</span><span class="o">=</span><span class="n">genes</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:root:1280 training genes are saved in `uns``training_genes` of both single cell and spatial Anndatas.
INFO:root:14785 overlapped genes are saved in `uns``overlap_genes` of both single cell and spatial Anndatas.
INFO:root:uniform based density prior is calculated and saved in `obs``uniform_density` of the spatial Anndata.
INFO:root:rna count based density prior is calculated and saved in `obs``rna_count_based_density` of the spatial Anndata.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[26]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Use the map_cells_to_space function from the tangram to map cells from &quot;adata_sc&quot; onto &quot;adata_st&quot;.</span>
<span class="c1"># The mapping use &quot;cells&quot; mode, which assign each cell from adata_sc to a location within the spatial transcriptomics space based on its gene expression profile.</span>
<span class="n">ad_map</span> <span class="o">=</span> <span class="n">tg</span><span class="o">.</span><span class="n">map_cells_to_space</span><span class="p">(</span>
    <span class="n">adata_sc</span><span class="p">,</span>
    <span class="n">adata_st</span><span class="p">,</span>
    <span class="n">mode</span><span class="o">=</span><span class="s2">&quot;cells&quot;</span><span class="p">,</span>
    <span class="c1"># target_count=adata_st.obs.cell_count.sum(),</span>
    <span class="c1"># density_prior=np.array(adata_st.obs.cell_count) / adata_st.obs.cell_count.sum(),</span>
    <span class="n">num_epochs</span><span class="o">=</span><span class="mi">1000</span><span class="p">,</span>
    <span class="n">device</span><span class="o">=</span><span class="s2">&quot;cpu&quot;</span><span class="p">,</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:root:Allocate tensors for mapping.
INFO:root:Begin training with 1280 genes and rna_count_based density_prior in cells mode...
INFO:root:Printing scores every 100 epochs.
</pre></div></div>
</div>
<div class="nboutput docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
Score: 0.613, KL reg: 0.001
Score: 0.733, KL reg: 0.000
Score: 0.736, KL reg: 0.000
Score: 0.737, KL reg: 0.000
Score: 0.737, KL reg: 0.000
Score: 0.737, KL reg: 0.000
Score: 0.737, KL reg: 0.000
Score: 0.737, KL reg: 0.000
Score: 0.737, KL reg: 0.000
Score: 0.738, KL reg: 0.000
</pre></div></div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:root:Saving results..
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="n">ad_map</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[27]:
</pre></div>
</div>
<div class="output_area docutils container">
<div class="highlight"><pre>
AnnData object with n_obs × n_vars = 21697 × 324
    obs: &#39;sample_name&#39;, &#39;organism&#39;, &#39;donor_sex&#39;, &#39;cell_class&#39;, &#39;cell_subclass&#39;, &#39;cell_cluster&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_mt&#39;, &#39;log1p_total_counts_mt&#39;, &#39;pct_counts_mt&#39;, &#39;n_counts&#39;
    var: &#39;in_tissue&#39;, &#39;array_row&#39;, &#39;array_col&#39;, &#39;n_genes_by_counts&#39;, &#39;log1p_n_genes_by_counts&#39;, &#39;total_counts&#39;, &#39;log1p_total_counts&#39;, &#39;pct_counts_in_top_50_genes&#39;, &#39;pct_counts_in_top_100_genes&#39;, &#39;pct_counts_in_top_200_genes&#39;, &#39;pct_counts_in_top_500_genes&#39;, &#39;total_counts_MT&#39;, &#39;log1p_total_counts_MT&#39;, &#39;pct_counts_MT&#39;, &#39;n_counts&#39;, &#39;leiden&#39;, &#39;cluster&#39;, &#39;uniform_density&#39;, &#39;rna_count_based_density&#39;
    uns: &#39;train_genes_df&#39;, &#39;training_history&#39;
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[28]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Project &quot;Cell_subclass&quot; annotations from a single-cell RNA sequencing (scRNA-seq) dataset onto a spatial transcriptomics dataset, based on a previously computed cell-to-space mapping</span>
<span class="n">tg</span><span class="o">.</span><span class="n">project_cell_annotations</span><span class="p">(</span><span class="n">ad_map</span><span class="p">,</span> <span class="n">adata_st</span><span class="p">,</span> <span class="n">annotation</span><span class="o">=</span><span class="s2">&quot;cell_subclass&quot;</span><span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area stderr docutils container">
<div class="highlight"><pre>
INFO:root:spatial prediction dataframe is saved in `obsm` `tangram_ct_pred` of the spatial AnnData.
</pre></div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Print adata_st.obsm[&#39;tangram_ct_pred&#39;]</span>
<span class="n">adata_st</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s1">&#39;tangram_ct_pred&#39;</span><span class="p">]</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[29]:
</pre></div>
</div>
<div class="output_area rendered_html docutils container">
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Pvalb</th>
      <th>L4</th>
      <th>Vip</th>
      <th>L2/3 IT</th>
      <th>Lamp5</th>
      <th>NP</th>
      <th>Sst</th>
      <th>L5 IT</th>
      <th>Oligo</th>
      <th>L6 CT</th>
      <th>...</th>
      <th>L5 PT</th>
      <th>Astro</th>
      <th>L6b</th>
      <th>Endo</th>
      <th>Peri</th>
      <th>Meis2</th>
      <th>Macrophage</th>
      <th>CR</th>
      <th>VLMC</th>
      <th>SMC</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>AAATGGCATGTCTTGT-1</th>
      <td>7.809727</td>
      <td>0.602690</td>
      <td>7.684955</td>
      <td>0.826228</td>
      <td>4.400264</td>
      <td>2.589142</td>
      <td>3.143504</td>
      <td>5.196132</td>
      <td>0.155253</td>
      <td>7.294017</td>
      <td>...</td>
      <td>8.664869</td>
      <td>2.573349</td>
      <td>0.281258</td>
      <td>0.412784</td>
      <td>0.000076</td>
      <td>0.000066</td>
      <td>1.247531</td>
      <td>0.070499</td>
      <td>0.167258</td>
      <td>0.627665</td>
    </tr>
    <tr>
      <th>AACAACTGGTAGTTGC-1</th>
      <td>5.715232</td>
      <td>0.000604</td>
      <td>14.581396</td>
      <td>0.207686</td>
      <td>5.556412</td>
      <td>3.897364</td>
      <td>3.344852</td>
      <td>10.164038</td>
      <td>0.450173</td>
      <td>2.114142</td>
      <td>...</td>
      <td>7.679641</td>
      <td>1.993195</td>
      <td>1.645181</td>
      <td>0.518714</td>
      <td>0.078318</td>
      <td>0.000459</td>
      <td>0.534635</td>
      <td>0.066145</td>
      <td>0.000130</td>
      <td>0.299213</td>
    </tr>
    <tr>
      <th>AACAGGAAATCGAATA-1</th>
      <td>4.831826</td>
      <td>1.001870</td>
      <td>7.364245</td>
      <td>1.041612</td>
      <td>6.277684</td>
      <td>0.413503</td>
      <td>8.242720</td>
      <td>6.679482</td>
      <td>0.496779</td>
      <td>14.989484</td>
      <td>...</td>
      <td>0.562375</td>
      <td>1.995712</td>
      <td>0.680008</td>
      <td>0.542730</td>
      <td>0.200446</td>
      <td>0.684655</td>
      <td>0.651862</td>
      <td>0.058122</td>
      <td>0.000165</td>
      <td>0.333420</td>
    </tr>
    <tr>
      <th>AACCCAGAGACGGAGA-1</th>
      <td>8.589804</td>
      <td>4.175763</td>
      <td>4.638950</td>
      <td>4.470548</td>
      <td>7.834509</td>
      <td>0.000336</td>
      <td>7.685515</td>
      <td>12.343997</td>
      <td>0.488403</td>
      <td>2.658836</td>
      <td>...</td>
      <td>0.001456</td>
      <td>2.909278</td>
      <td>0.000471</td>
      <td>0.560556</td>
      <td>0.000042</td>
      <td>0.336403</td>
      <td>0.900331</td>
      <td>0.000066</td>
      <td>0.557967</td>
      <td>0.550292</td>
    </tr>
    <tr>
      <th>AACCGTTGTGTTTGCT-1</th>
      <td>9.990313</td>
      <td>5.441794</td>
      <td>4.714663</td>
      <td>1.678014</td>
      <td>5.508845</td>
      <td>1.483209</td>
      <td>6.444257</td>
      <td>14.175096</td>
      <td>1.404791</td>
      <td>0.000503</td>
      <td>...</td>
      <td>1.804151</td>
      <td>1.123749</td>
      <td>0.998521</td>
      <td>1.451936</td>
      <td>0.056683</td>
      <td>0.000208</td>
      <td>0.219766</td>
      <td>0.034044</td>
      <td>0.000118</td>
      <td>0.581478</td>
    </tr>
    <tr>
      <th>...</th>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
      <td>...</td>
    </tr>
    <tr>
      <th>TTGGATTGGGTACCAC-1</th>
      <td>5.641819</td>
      <td>2.058308</td>
      <td>12.881126</td>
      <td>1.627751</td>
      <td>3.359651</td>
      <td>1.249576</td>
      <td>13.079297</td>
      <td>9.607204</td>
      <td>0.419696</td>
      <td>0.616700</td>
      <td>...</td>
      <td>6.947644</td>
      <td>2.559545</td>
      <td>0.000223</td>
      <td>0.490524</td>
      <td>0.019175</td>
      <td>0.000077</td>
      <td>0.261640</td>
      <td>0.000564</td>
      <td>0.250397</td>
      <td>0.716284</td>
    </tr>
    <tr>
      <th>TTGGCTCGCATGAGAC-1</th>
      <td>3.322495</td>
      <td>4.845051</td>
      <td>4.550827</td>
      <td>10.971983</td>
      <td>7.764623</td>
      <td>0.000582</td>
      <td>12.342369</td>
      <td>9.450901</td>
      <td>0.159477</td>
      <td>0.092170</td>
      <td>...</td>
      <td>0.000585</td>
      <td>0.957426</td>
      <td>0.001045</td>
      <td>0.404693</td>
      <td>0.006123</td>
      <td>0.036123</td>
      <td>0.290322</td>
      <td>0.000344</td>
      <td>0.237261</td>
      <td>0.442137</td>
    </tr>
    <tr>
      <th>TTGTATCACACAGAAT-1</th>
      <td>3.573624</td>
      <td>0.001394</td>
      <td>7.127228</td>
      <td>0.003228</td>
      <td>4.591670</td>
      <td>4.277098</td>
      <td>9.285900</td>
      <td>9.232115</td>
      <td>0.697667</td>
      <td>6.629681</td>
      <td>...</td>
      <td>4.475281</td>
      <td>0.783166</td>
      <td>0.737044</td>
      <td>0.343111</td>
      <td>0.067538</td>
      <td>0.000370</td>
      <td>0.167680</td>
      <td>0.040888</td>
      <td>0.239439</td>
      <td>0.220072</td>
    </tr>
    <tr>
      <th>TTGTGGCCCTGACAGT-1</th>
      <td>9.355780</td>
      <td>1.569948</td>
      <td>9.292291</td>
      <td>0.141470</td>
      <td>0.208387</td>
      <td>0.386280</td>
      <td>5.018963</td>
      <td>4.222025</td>
      <td>0.805368</td>
      <td>6.953451</td>
      <td>...</td>
      <td>3.673550</td>
      <td>2.167665</td>
      <td>1.065244</td>
      <td>0.610526</td>
      <td>0.061187</td>
      <td>0.000065</td>
      <td>0.723517</td>
      <td>0.000139</td>
      <td>0.355459</td>
      <td>0.213958</td>
    </tr>
    <tr>
      <th>TTGTTAGCAAATTCGA-1</th>
      <td>3.873220</td>
      <td>20.980103</td>
      <td>6.624079</td>
      <td>1.769871</td>
      <td>1.144407</td>
      <td>0.000994</td>
      <td>5.585440</td>
      <td>14.799476</td>
      <td>0.827283</td>
      <td>0.002911</td>
      <td>...</td>
      <td>0.954696</td>
      <td>1.307711</td>
      <td>0.000311</td>
      <td>0.545261</td>
      <td>0.092780</td>
      <td>0.013870</td>
      <td>0.106881</td>
      <td>0.000795</td>
      <td>0.183378</td>
      <td>0.195309</td>
    </tr>
  </tbody>
</table>
<p>324 rows × 23 columns</p>
</div></div>
</div>
<div class="nbinput docutils container">
<div class="prompt highlight-none notranslate"><div class="highlight"><pre><span></span>[30]:
</pre></div>
</div>
<div class="input_area highlight-ipython3 notranslate"><div class="highlight"><pre><span></span><span class="c1"># Concatenate the predicted cell type labels computed by the tangram during the cell-to-space mapping step to &#39;adata_st.obs&#39;</span>
<span class="n">adata_st</span><span class="o">.</span><span class="n">obs</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">concat</span><span class="p">([</span><span class="n">adata_st</span><span class="o">.</span><span class="n">obs</span><span class="p">,</span> <span class="n">adata_st</span><span class="o">.</span><span class="n">obsm</span><span class="p">[</span><span class="s2">&quot;tangram_ct_pred&quot;</span><span class="p">]],</span> <span class="n">axis</span><span class="o">=</span><span class="mi">1</span><span class="p">)</span>

<span class="c1"># Create a spatial scatter plot showing the distribution of different cell types</span>
<span class="n">sq</span><span class="o">.</span><span class="n">pl</span><span class="o">.</span><span class="n">spatial_scatter</span><span class="p">(</span>
    <span class="n">adata_st</span><span class="p">,</span>
    <span class="n">color</span><span class="o">=</span><span class="p">[</span><span class="s2">&quot;L2/3 IT&quot;</span><span class="p">,</span> <span class="s2">&quot;L4&quot;</span><span class="p">,</span> <span class="s2">&quot;L5 IT&quot;</span><span class="p">,</span> <span class="s2">&quot;L5 PT&quot;</span><span class="p">,</span> <span class="s2">&quot;L6 CT&quot;</span><span class="p">,</span> <span class="s2">&quot;L6 IT&quot;</span><span class="p">,</span> <span class="s2">&quot;L6b&quot;</span><span class="p">],</span>
<span class="p">)</span>
</pre></div>
</div>
</div>
<div class="nboutput nblast docutils container">
<div class="prompt empty docutils container">
</div>
<div class="output_area docutils container">
<img alt="../_images/Deconvolution_Reproducibility_with_original_data_29_0.png" src="../_images/Deconvolution_Reproducibility_with_original_data_29_0.png" />
</div>
</div>
</section>
</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="页脚">
        <a href="Installation.html" class="btn btn-neutral float-left" title="Installation" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> 上一页</a>
        <a href="Application%20with%20new%20data.html" class="btn btn-neutral float-right" title="Application with new data" accesskey="n" rel="next">下一页 <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; 版权所有 2023, LINSENLIN.</p>
  </div>

  利用 <a href="https://www.sphinx-doc.org/">Sphinx</a> 构建，使用的 
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">主题</a>
    由 <a href="https://readthedocs.org">Read the Docs</a> 开发.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>